### PR TITLE
[Fix] Account resident weight memory in KV sizing

### PR DESCRIPTION
**Pull request by galletas1712**
## Motivation

The IPC weight-cache daemon keeps model weights resident before the engine
starts, so `ModelRunner` samples its pre-load free-memory baseline after those
weights have already reduced free memory. Without correcting that baseline, KV
sizing reserves headroom from only the remaining free memory and can
over-allocate the KV cache.

The correction must be rank-local before the distributed `MIN`: adding
preloaded weight memory after ranks have already reduced their baselines is
incorrect when rank footprints differ. Target KV sizing also happens after
draft loading, so target, multi-runner, and future speculative draft weights
must be aggregated before sizing.

## Modifications

- Add a generic rank-local `BaseModelLoader.preloaded_weights_bytes` contract,
  defaulting to zero, and validate loader-provided values before use.
- Aggregate preloaded bytes across target, multi-runner, and speculative draft
  workers. The current built-in IPC weight cache plus speculative decoding
  combination remains explicitly rejected because draft-model daemon support
  is not implemented yet, but the accounting path is ready for it.
- Immediately before target KV-pool allocation, compute
  `MIN(local pre-load baseline + local preloaded weight bytes)` and update
  `pre_model_load_memory`, so initial sizing and post-capture resizing use the
  same corrected value.
- Measure the daemon's process-local PyTorch allocator growth during model
  loading and return it through the IPC protocol as
  `preloaded_weights_bytes`.

## Accuracy Tests

Ran a real persistent-daemon restart A/B with
`Qwen/Qwen2.5-32B-Instruct` in BF16 at TP=1 on an 80 GB GPU. The standalone
daemon retained 63,004 MiB of GPU memory while two independent engine clients
were started sequentially against the same daemon.

| Result | Without this fix | With this fix |
|---|---:|---:|
| Client pre-load available memory | 17.23 GB | 17.23 GB |
| Reported preloaded weights | Not available | 65,546,485,760 bytes (61.045 GiB) |
| KV cache tokens | 63,434 | 38,430 |
| K + V allocation | 15.48 GB | 9.38 GB |
| Memory before decode graph capture | 0.88 GB | 4.37 GB |
| First client start | CUDA OOM during graph capture | Healthy; inference succeeded |
| Second client start | Same CUDA OOM | Healthy; inference succeeded |

Without the correction, both starts selected the same oversized KV cache and
failed during decode CUDA graph capture while the daemon remained alive. With
the correction, both starts completed graph capture, reached health, and served
a completion against the same persistent daemon.

This validates the final implementation and reproduces the accounting defect
as a deterministic startup failure with a model whose resident weights occupy
more than half of GPU memory.

## Speed Tests and Profiling

Not benchmarked. The steady-state inference path is unchanged. Startup adds
two process-local allocator reads and one existing-world-group reduction before
KV-pool allocation.

## Validation

- Large-model persistent-daemon restart A/B passed as described above
- Python syntax checks passed for all changed production files
- Ruff error checks passed for all changed production files
- Full Ruff checks passed for the central accounting and IPC modules
- Diff whitespace checks passed

## Checklist

- [x] Format code according to the repository hooks.
- [x] Validate the exact final implementation with a persistent GPU daemon and
      two engine starts.
- [ ] Update documentation — not applicable; no user-facing option or workflow
      changed.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #33039299218](https://github.com/sgl-project/sglang/actions/runs/33039299218)
Latest PR Test (Extra): :x: [Run #33039298996](https://github.com/sgl-project/sglang/actions/runs/33039298996)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33039299152](https://github.com/sgl-project/sglang/actions/runs/33039299152)

### CODE REVIEW

**Review by alexnails — COMMENTED** [review]
quick comments:
1. we only need this during weight daemon enabled
2. reduce_min_gpu_memory and its result local_prep_model_load feels like something that can exist within global_ctx, or even be part of available_gpu_memory, can you talk more about your thoughts on this?

**Comment by galletas1712** [comment]
Thanks for the feedback @alexnails 

I adjusted the PR based on your feedback

1. Agreed the correction is only needed when weights are already resident (the daemon). account_preloaded_weights is a no-op when the loader reports 0 bytes, so a normal start doesn’t do another reduction. I left it as a loader field defaulting to 0 rather than an explicit --weight-cache-mode check, so any loader that maps already-resident weights can use the same hook.

2. I think the PR was originally a bit over-defensive. The daemon only breaks the first measurement of pre_model_load_memory so I just got rid of reduce_min_gpu_memory and local_model_pre_load_memory. I didn't fold anything into global context or available_gpu_memory though because by the time we call available_gpu_memory during KV cache allocation, it's too late.

File: python/sglang/srt/model_executor/model_runner.py [hunk 1]
```diff
@@ -822,6 +822,24 @@ def max_decode_logits_rows(self) -> int:
         capture_bs, _ = get_batch_sizes_to_capture(self, num_tokens_per_req)
         return max(capture_bs) * num_tokens_per_req
 
+    @property
+    def preloaded_weights_bytes(self) -> int:
+        value = getattr(self.loader, "preloaded_weights_bytes", 0)
```
Comment (liusy58) [inline python/sglang/srt/model_executor/model_runner.py]: We had better try to avoid using `getattr`

**Review by liusy58 — APPROVED** [review]

File: python/sglang/srt/model_executor/model_runner.py (hunk 1 above)
Comment (galletas1712) [reply python/sglang/srt/model_executor/model_runner.py]: Thanks for the catch - fixed.
